src/player.cpp

```diff
@@ -11158,162 +11158,184 @@ void player::gunmod_add( item &gun, item &mod )
+        return INT_MAX;
     }
-
-    vehicle *veh = g->m.veh_at( pos() );
-    if( veh != nullptr && veh->player_in_control(*this) ) {
-        add_msg(m_info, _("It's a bad idea to read while driving!"));
-        return false;
+    const skill_id &skill = type->skill;
+    const auto &skill_level = get_skill_level( skill );
+    if( skill && skill_level < type->req ) {
+        reasons.push_back( string_format( _( "You don't know enough about %s to understand the jargon!" ),
+                                          skill.obj().name().c_str() ) );
+        return INT_MAX;
+    }
+
+    // Check for conditions that would disqualify us if no NPCs can read to us
+    if( fine_detail_vision_mod() > 4 ) {
+        reasons.push_back( _( "You can't see to read!" ) );
+    } else if( type->intel > 0 && has_trait( "ILLITERATE" ) ) {
+        reasons.push_back( _( "You're illiterate!" ) );
+    } else if( has_trait( "HYPEROPIC" ) && !is_wearing( "glasses_reading" ) &&
+               !is_wearing( "glasses_bifocal" ) && !has_effect( effect_contacts ) ) {
+        reasons.push_back( _( "Your eyes won't focus without reading glasses." ) );
+    } else {
+        reader = this;
+    }
+
+    //Check for NPCs to read for you, negates Illiterate, Far Sighted, and darkness
+    int time_taken = INT_MAX;
+    if( !reader ) {
+        if( is_deaf() ) {
+            reasons.push_back( _( "Maybe someone could read that to you, but you're deaf!" ) );
+            return INT_MAX;
+        }
+        for( const npc *elem : g->active_npc ) {
+            // Check for disqualifying factors:
+            if( !elem->is_friend() ) {
+                ; // Don't consider non-friends
+            } else if( rl_dist( pos(), elem->pos() ) >= PICKUP_RANGE ) {
+                if( sees( *elem ) ) {
+                    reasons.push_back( string_format( _( "Maybe %s could read that to you, but they're too far away." ),
+                                                      elem->disp_name().c_str() ) );
+                }
+            } else if( elem->in_sleep_state() ) {
+                if( sees( *elem ) ) {
+                    reasons.push_back( string_format( _( "Maybe %s could read that to you, but they're sleeping." ),
+                                                      elem->disp_name().c_str() ) );
+                }
+            } else if( !elem->sees( *this ) ) {
+                if( sees( *elem ) ) {
+                    reasons.push_back( string_format( _( "Maybe %s could read that to you, but they can't see you." ),
+                                                      elem->disp_name().c_str() ) );
+                    }
+            } else if( type->intel > 0 && elem->has_trait( "ILLITERATE" ) ) {
+                reasons.push_back( string_format( _( "%s says: \"I'd read that to you, but I'm illiterate!\"" ),
+                                                  elem->disp_name().c_str() ) );
```

Comment:
Would be better to just say "npc is illiterate/has bad eyesight etc." and move it up. In general, states like "won't ever help" should be printed before "current state prevents the NPC from helping".